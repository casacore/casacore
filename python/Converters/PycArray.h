//# PycArray.h: Class to convert an Array to/from Python
//# Copyright (C) 2006
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: PycArray.h,v 1.4 2006/11/06 00:14:44 gvandiep Exp $


#ifndef PYRAP_PYCARRAY_H
#define PYRAP_PYCARRAY_H

//# Includes
// include first to avoid _POSIX_C_SOURCE redefined warnings
#include <boost/python.hpp>
#include <boost/python/object.hpp>
#include <casacore/casa/Arrays/Array.h>
#include <casacore/casa/Containers/ValueHolder.h>
#include <casacore/casa/Utilities/DataType.h>
#include <casacore/casa/Exceptions/Error.h>
#include <iostream>

namespace casa { namespace python {


  // <summary>
  // A class to convert an Array to/from Python objects.
  // </summary>

  // <use visibility=export>
  // <reviewed reviewer="" date="" tests="">
  // </reviewed>

  // <synopsis>
  // </synopsis>

  // Check if the PyObject is an array object.
  Bool PycArrayCheck (PyObject* obj_ptr);

  // Check if the PyObject is an array scalar object.
  Bool PycArrayScalarCheck (PyObject* obj_ptr);

  // Get the data type of the array scalar object.
  // It returns TpBool, TpInt, TpFloat, or TpComplex.
  // TpOther is returned if unrecognized.
  DataType PycArrayScalarType (PyObject* obj_ptr);

  struct casa_array_from_python
  {
    // Constructs an Array from a Python object.
    // If copyData=False, the array data is only copied if needed meaning
    // that the Array object in the ValueHolder can reference the data in
    // Python array.
    // That should only be used if the ValueHolder and its Array will be
    // destructed before the Python array.
    static ValueHolder makeArray(PyObject* obj_ptr, Bool copyData=False);

    // Construct an Array<String> from a special Python dict object.
    static ValueHolder makeArrayFromDict (PyObject* obj_ptr);

    // Construct a scalar from an array scalar (i.e. element in array).
    static ValueHolder makeScalar (PyObject* obj_ptr);
  };

  // Do the actual making of the PyArrayObject.
  // Specialize for strings.
  // <group>
  template <typename T>
  boost::python::object makePyArrayObject (casa::Array<T> const& arr);
  template <>
  boost::python::object makePyArrayObject (casa::Array<String> const& arr);
  // </group>

  // Convert Array to Python.
  template <typename T>
  struct casa_array_to_python
  {
    static boost::python::object makeobject (Array<T> const& arr)
      { return makePyArrayObject (arr); }
    static PyObject* convert (Array<T> const& c)
      { return boost::python::incref(makeobject(c).ptr()); }
  };

}}

#endif
