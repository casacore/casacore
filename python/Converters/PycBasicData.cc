//# PycBasicData.cc: Convert casa data types to/from python
//# Copyright (C) 2006
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: PycBasicData.cc,v 1.4 2007/01/29 04:23:01 mmarquar Exp $

#include <python/Converters/PycBasicData.h>
#include <python/Converters/PycArrayNP.h>

using namespace boost::python;


namespace casa { namespace python {

  std::map<std::string,bool> pyregistry::_registry;
  bool pyregistry::get (const std::string& name)
  {
    return _registry[name];
  }
  void pyregistry::set (const std::string& name)
  {
    _registry[name] = true;
  }


  void convert_casa_string::reg()
  {
    std::string tname(typeid(casa::String).name());
    if (! pyregistry::get (tname)) {
      pyregistry::set (tname);
      boost::python::to_python_converter<String, casa_string_to_python_str>();
      casa_string_from_python_str();
    }
  }

  void convert_casa_iposition::reg()
  {
    std::string tname(typeid(casa::IPosition).name());
    if (! pyregistry::get (tname)) {
      pyregistry::set (tname);
      casa_iposition_to_list();
      from_python_sequence < casa::IPosition,
                             casa_reversed_variable_capacity_policy > ();
    }
  }


  void register_convert_basicdata()
  {
    casa::python::numpy::register_convert_arrayscalars();
    casa::python::register_convert_casa_string();
    casa::python::register_convert_casa_iposition();
    casa::python::register_convert_casa_vector<casa::Bool>();
    casa::python::register_convert_casa_vector<casa::Int>();
    casa::python::register_convert_casa_vector<casa::Double>();
    casa::python::register_convert_casa_vector<casa::Float>();
    casa::python::register_convert_casa_vector<casa::DComplex>();
    casa::python::register_convert_casa_vector<casa::String>();
  }


  bool getSeqObject (object& py_obj)
  {
    // Restriction to list, tuple, iter, xrange until
    // Boost.Python overload resolution is enhanced.
    //  PySequence_Check() is used for numarray.
    PyObject* obj_ptr = py_obj.ptr();
    if (!(PyList_Check(obj_ptr)
	  || PyTuple_Check(obj_ptr)
	  || PyIter_Check(obj_ptr)
	  || PyRange_Check(obj_ptr)
	  || PySequence_Check(obj_ptr) )) return false;
    // Must be a measurable sequence.
    int obj_size = -1;
    bool done = false;
    // Try to get attribute size, because length fails for a numpy scalar.
    try {
      // A numpy scalar size should be 1.
      object py_tmp = py_obj.attr("size");
      if (extract<int>(py_tmp) == 1) {
	done = true;
      }
    } catch (...) {
      PyErr_Clear();
    }
    // If it failed, try to get the length.
    if (!done) {
      obj_size = PyObject_Length(obj_ptr);
      if (obj_size < 0) {
	done = true;
	PyErr_Clear();
      }
    }
    // If we seem to have a numpy/numarray scalar, try to flatten it.
    // Return the flattened object.
    if (done) {
      done = false;
      object py_flat;
      // Try if the object is a scalar numarray/numpy object which
      // can be flattened to a vector num object.
      try {
	py_flat = py_obj.attr("flatten")();    // numpy attr name
	done = true;
      } catch (...) {
	PyErr_Clear();
      }
      if (!done) {
	try {
	  py_flat = py_obj.attr("flat");       // numarray attr name
	  done = true;
	} catch (...) {
	  PyErr_Clear();
	}
      }
      if (done) py_obj = py_flat;
    }
    // If it failed, try to get the length.
    if (!done) {
      obj_size = PyObject_Length(obj_ptr);
      if (obj_size >= 0) {
	done = true;
      } else {
	PyErr_Clear();
      }
    }
    return done;
  }


}}
