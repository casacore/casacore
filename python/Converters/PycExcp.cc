//# PycExcp.cc: Functions to convert a C++ exception to Python
//# Copyright (C) 2006
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: PycExcp.cc,v 1.1 2006/10/17 03:33:50 gvandiep Exp $

#include <casacore/python/Converters/PycExcp.h>
#include <casacore/casa/Containers/IterError.h>
//# The following include is necessary to work around a Boost-Python problem.
#ifndef PYRAP_NO_BOOSTPYTHON_FIX
# include <boost/type_traits/add_reference.hpp>
#endif
#include <boost/python/exception_translator.hpp>

namespace casacore { namespace python {

  void translate_iterexcp (const casacore::IterError& e)
  {
    // Use the Python 'C' API to set up an exception object
    PyErr_SetString(PyExc_StopIteration, e.what());
  }

  void translate_stdexcp (const std::exception& e)
  {
    // Use the Python 'C' API to set up an exception object
    PyErr_SetString(PyExc_RuntimeError, e.what());
  }


  //# Note that the most general exception must be registered first.
  void register_convert_excp()
  {
    boost::python::register_exception_translator<std::exception>
      (&translate_stdexcp);
    boost::python::register_exception_translator<casacore::IterError>
      (&translate_iterexcp);
  }

}}
