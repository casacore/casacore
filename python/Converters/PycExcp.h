//# PycExcp.h: Functions to convert a C++ exception to Python
//# Copyright (C) 2006
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: PycExcp.h,v 1.3 2006/10/17 03:33:50 gvandiep Exp $

#ifndef PYRAP_PYCEXCP_H
#define PYRAP_PYCEXCP_H

#include <casacore/casa/Containers/IterError.h>

namespace casacore { namespace python {

  // Convert an IterError exception to a Python StopIteration.
  // In this way an iteration loop can be done.
  void translate_iterexcp (const casacore::IterError& e);

  // Register exception translators for casacore::IterError.
  void register_convert_excp();

}}

#endif
