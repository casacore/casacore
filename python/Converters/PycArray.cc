//# PycArray.cc: Class to convert an Array to/from Python
//# Copyright (C) 2006
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: PycArray.cc,v 1.4 2006/11/06 00:14:44 gvandiep Exp $

#include <python/Converters/PycArray.tcc>
#include <python/Converters/PycArrayNP.h>
#include <python/Converters/PycBasicData.h>
#include <casa/Arrays/ArrayMath.h>
#include <boost/python/dict.hpp>
#include <sysmodule.h>

using namespace boost::python;

namespace casa { namespace python {

  Bool PycArrayCheck (PyObject* obj_ptr)
  {
    return numpy::PycArrayCheck(obj_ptr);
  }

  Bool PycArrayScalarCheck (PyObject* obj_ptr)
  {
    int type;
    return numpy::PycArrayScalarCheck(obj_ptr, type);
  }

  DataType PycArrayScalarType (PyObject* obj_ptr)
  {
    return numpy::PycArrayScalarType(obj_ptr);
  }

  ValueHolder casa_array_from_python::makeArray (PyObject* obj_ptr,
						 Bool copyData)
  {
    if (! numpy::PycArrayCheck(obj_ptr)) {
      throw AipsError ("PycArray: python object is not a numpy array");
    }
    return numpy::makeArray (obj_ptr, copyData);
  }

  ValueHolder casa_array_from_python::makeScalar (PyObject* obj_ptr)
  {
    int type;
    if (!numpy::PycArrayScalarCheck(obj_ptr, type)) {
      throw AipsError ("PycArray: python object is not a numpy array scalar");
    }
    return numpy::makeScalar (obj_ptr, type);
  }

  ValueHolder casa_array_from_python::makeArrayFromDict (PyObject* obj_ptr)
  {
    if (! PyDict_Check(obj_ptr)) {
      throw AipsError ("PycArray: python object is not a dict");
    }
    dict d = extract<dict>(obj_ptr)();
    IPosition shp = extract<IPosition>(d.get("shape").ptr())();
    Array<String> arr = extract<Vector<String> >(d.get("array").ptr())();
    if (Int(arr.size()) != shp.product()) {
      throw AipsError("PycArray: array size mismatches the shape");
    }
    return ValueHolder(arr.reform (shp));
  }

  template <>
  object makePyArrayObject (casa::Array<String> const& arr)
  {
    object a = to_list< Array<String> >::makeobject (arr);
    if (arr.ndim() == 1) {
      return a;
    }
    dict d;
    d.setdefault (std::string("shape"),
		  to_list<IPosition>::makeobject (arr.shape()));
    d.setdefault (std::string("array"), a);
    return d;
  }


  // Instantiate the templates.
  template boost::python::object makePyArrayObject
    (casa::Array<Bool> const& arr);
  template boost::python::object makePyArrayObject
    (casa::Array<uChar> const& arr);
  template boost::python::object makePyArrayObject
    (casa::Array<Short> const& arr);
  template boost::python::object makePyArrayObject
    (casa::Array<uShort> const& arr);
  template boost::python::object makePyArrayObject
    (casa::Array<Int> const& arr);
  template boost::python::object makePyArrayObject
    (casa::Array<uInt> const& arr);
  template boost::python::object makePyArrayObject
    (casa::Array<Int64> const& arr);
  template boost::python::object makePyArrayObject
    (casa::Array<Float> const& arr);
  template boost::python::object makePyArrayObject
    (casa::Array<Double> const& arr);
  template boost::python::object makePyArrayObject
    (casa::Array<Complex> const& arr);
  template boost::python::object makePyArrayObject
    (casa::Array<DComplex> const& arr);

}}
