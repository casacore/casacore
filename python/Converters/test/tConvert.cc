//# tConvert.cc: Test program for libpython's C++/Python converters
//# Copyright (C) 2006
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: tConvert.cc,v 1.4 2006/11/06 00:14:44 gvandiep Exp $

#include <python/Converters/PycExcp.h>
#include <python/Converters/PycBasicData.h>
#include <python/Converters/PycValueHolder.h>
#include <python/Converters/PycRecord.h>
#include <python/Converters/PycArray.h>
#include <casa/Arrays/ArrayIO.h>

#include <boost/python.hpp>

using namespace boost::python;

namespace casa { namespace python {

  template<typename T>
  std::ostream& operator<< (std::ostream& os, const std::vector<T>& vec)
  {
    os << '[';
    for (uInt i=0; i<vec.size(); ++i) {
      if (i > 0) os << ", ";
      os << "vecuInt " << vec[i];
    }
    os << ']';
    return os;
  }

  struct TConvert
  {
    TConvert() {}
    Bool  testbool (Bool in)
      {cout << "bool " << in << endl; return in;}
    Int testint (Int in)
      {cout << "Int " << in << endl; return in;}
    Int64 testint64 (Int64 in)
      {cout << "Int64 " << in << endl; return in;}
    Int testssize (::ssize_t in)
      {cout << "ssize " << in << endl; return in;}
    Float testfloat (Float in)
      {cout << "Float " << in << endl; return in;}
    Double testdouble (Double in)
      {cout << "Double " << in << endl; return in;}
    Complex testcomplex (const Complex& in)
      {cout << "Complex " << in << endl; return in;}
    DComplex testdcomplex (const DComplex& in)
      {cout << "DComplex " << in << endl; return in;}
    String teststring (const String& in)
      {cout << "String " << in << endl; String out=in; return out;}
    Record testrecord (const Record& in)
      {cout << "Record "; in.print(cout); cout << endl; return in;}
    ValueHolder testvh (const ValueHolder& in)
      {cout << "VH " << in.dataType() << endl; return in;}
    Vector<Int> testvecint (const Vector<int>& in)
      {cout << "VecInt " << in << endl; return in;}
    Vector<DComplex> testveccomplex (const Vector<DComplex>& in)
      {cout << "VecComplex " << in << endl; return in;}
    Vector<String> testvecstr (const Vector<String>& in)
      {cout << "VecStr " << in << endl; return in;}
    std::vector<uInt> teststdvecuint (const std::vector<uInt>& in)
      {cout << "vecuInt " << in << endl; return in;}
    std::vector<std::vector<uInt> > teststdvecvecuint
    (const std::vector<std::vector<uInt> >& in)
      {cout << "vecvecuInt " << in << endl; return in;}
    IPosition testipos (const IPosition& in)
      {cout << "IPos " << in << endl; return in;}
    Bool canusenumpy()
      {return PycCanUseNumpy();}
    Bool canusenumarray()
      {return PycCanUseNumarray();}
  };


  void testConvert()
  {
    class_<TConvert> ("tConvert", init<>())
      .def ("testbool",       &TConvert::testbool)
      .def ("testint",        &TConvert::testint)
      .def ("testint64",      &TConvert::testint64)
      .def ("testssize",      &TConvert::testssize)
      .def ("testfloat",      &TConvert::testfloat)
      .def ("testdouble",     &TConvert::testdouble)
      .def ("testcomplex",    &TConvert::testcomplex)
      .def ("testdcomplex",   &TConvert::testdcomplex)
      .def ("teststring",     &TConvert::teststring)
      .def ("testrecord",     &TConvert::testrecord)
      .def ("testvh",         &TConvert::testvh)
      .def ("testvecint",     &TConvert::testvecint)
      .def ("testveccomplex", &TConvert::testveccomplex)
      .def ("testvecstr",     &TConvert::testvecstr)
      .def ("teststdvecuint", &TConvert::teststdvecuint)
      .def ("teststdvecvecuint", &TConvert::teststdvecvecuint)
      .def ("testipos",       &TConvert::testipos)
      .def ("canusenumpy",    &TConvert::canusenumpy)
      .def ("canusenumarray", &TConvert::canusenumarray)
      ;
  }

}}


BOOST_PYTHON_MODULE(_tConvert)
{
  // Register the required converters.
  casa::python::register_convert_excp();
  casa::python::register_convert_basicdata();
  casa::python::register_convert_casa_valueholder();
  casa::python::register_convert_casa_record();
  casa::python::register_convert_std_vector<casa::uInt>();
  casa::python::register_convert_std_vector<std::vector<casa::uInt> >();

  // Execute the test.
  casa::python::testConvert();
}
