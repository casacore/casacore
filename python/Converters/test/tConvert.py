#!/usr/bin/env python
from _tConvert import *

def dotest(t):

    print ''
    print 'begin dotest'
    print t.testbool (True);
    print t.testbool (False);
    print t.testint (-1);
    print t.testint (10L);
    print t.testint64 (-123456789013L);
    print t.testint64 (123456789014L);
    print t.testssize (-2);
    print t.testssize (11);
    print t.testfloat (3.14);
    print t.testfloat (12);
    print t.teststring ("this is a string");

    print t.testipos ([2,3,4]);
    print t.testipos (1);
    print t.testipos (NUM.array([2]));
    print t.testipos (NUM.array(3));

    print t.testvecint ([1,2,3,4]);
    print t.testvecint ([]);
    print t.testvecint ((-1,-2,-3,-4));
    print t.testvecint (-10);
    print t.testvecint (NUM.array((10,11,12)));
    print t.testvecint (NUM.array(1));
    print t.testveccomplex ([1+2j, -1-3j, -1.5+2.5j]);
    print t.testvecstr (["a1","a2","b1","b2"])
    print t.testvecstr (())
    print t.testvecstr ("sc1")
    print t.teststdvecuint ([1,2,4])
    print t.teststdvecuint (())
    print t.teststdvecuint (10)
    print t.teststdvecvecuint ([[1,2,4]])
    print t.teststdvecvecuint ((()))
    print t.teststdvecvecuint (())
    print t.teststdvecvecuint ([1,2,4])
    print t.teststdvecvecuint (20)

    print t.testvh (True);
    print t.testvh (2);
    print t.testvh (1234567890123L);
    print t.testvh (1.3);
    print t.testvh (10-11j);
    print t.testvh ("str");
    print t.testvh ([True]) + 0;         # add 0 to convert numpy to integer
    print t.testvh ([2,4,6,8,10]);
    print t.testvh ([1.3,4,5,6]);
    print t.testvh ([10-11j,1+2j]);
#    print t.testvh ([]);
    print t.testvh (["str1","str2"]);
    print t.testvh ({"shape":[2,2],"array":["str1","str2","str3","str4"]});
    a  =  t.testvh ({"shape":[2,2],"array":["str1","str2","str3","str4"]});
    print a;
    print t.testvh (a);

    a  =  t.testvh ([10-11j,1+2j]);
    print a.shape;
    print t.testvh (a);

    b  =  NUM.int32([[2,3],[4,5]]);
    print b;
    print t.testvh (b);

    b  =  NUM.int32([1,2,3,4,5,6,7,8,9,10]);
    print b[2:9:2];
    print t.testvh (b[2:9:2]);

    b  =  NUM.array([1,2,3,4,5,6,7,8,9,10.]);
    print b[2:9:2];
    print t.testvh (b[2:9:2]);
    a = b[2:9:2];
    print t.testvh (a);

    print t.testvh(NUM.array([20.+10j]));
    print t.testvh(NUM.array([21.]));
    print t.testvh(NUM.array(21.));

    print '>>>';
    res = t.testvh (NUM.array([]));
    print '<<<';
    print res.shape;
    print '>>>';
    res = t.testvh (NUM.array([[]]));
    print '<<<';
    print res.shape;

    # On 64-bit machines the output also contains 'dtype=int32'
    # So leave it out.
    a = t.testrecord({"int":1, "int64":123456789012L, "str":"bc", 'vecint':[1,2,3]})
    print '>>>'
    print a
    print '<<<'
    print 'end dotest'
    print ''


def testarrvh(arr):
    print '    testarrvh';
    print t.testvh(arr);
    print t.testvh(arr[0]);
    print t.testvh([arr[0]]);
    print t.testvh([arr[0], arr[1]]);

def testarrb(arr):
    testarrvh(arr);
    print t.testbool(arr[0]);

def testarri(arr):
    testarrvh(arr);
    print t.testint(arr[0]);
    print t.testint64(arr[0]);
    print t.testssize(arr[0]);
    print t.testfloat(arr[0]);
    print t.testcomplex(arr[0]);

def testarrf(arr):
    testarrvh(arr);
    print t.testfloat(arr[0]);
    print t.testcomplex(arr[0]);

def testarrc(arr):
    testarrvh(arr);
    print t.testcomplex(arr[0]);

def testnps():
    testarrb(NUM.array([True,False]));
    testarri(NUM.int8([-6,-7]));
    testarri(NUM.uint8([5,6]));
    testarri(NUM.int16([-16,-17]));
    testarri(NUM.uint16([15,16]));
    testarri(NUM.int32([-26,-27]));
    testarri(NUM.uint32([25,26]));
    testarri(NUM.int64([-36,-37]));
    testarri(NUM.uint64([35,36]));
    testarrf(NUM.float32([-46,-47]));
    testarrf(NUM.float64([45,46]));
    testarrc(NUM.complex64([-56-66j,-57-67j]));
    testarrc(NUM.complex128([-76-86j,-77-87j]));

def testnp():
    # Test byte and sbyte.
    b = NUM.int8([-1,-2]);
    print t.testvh(b);
    b = NUM.uint8([211,212]);
    print t.testvh(b);
    print '>>>';
    res = t.testvh(NUM.array([]));
    print '<<<';
    print res.shape;
    print t.testvh(NUM.array([["abcd","c"],["12","x12"]]));
    testnps();

if __name__ == "__main__":

    import numpy as NUM;
    t = tConvert();
    print "Doing numpy/array test ..."
    testnp();
    # Do other tests.
    dotest(t)
