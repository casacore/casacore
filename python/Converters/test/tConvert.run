#!/bin/sh

# Use .run file for 2 reasons:
#  1. tConvert.py is not executable
#  2. Do not use valgrind on it

python tConvert.py
