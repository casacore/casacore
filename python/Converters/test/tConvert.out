Doing numpy/array test ...
VH Array<Short>
[-1 -2]
VH Array<Short>
[211 212]
>>>
VH Array<double>
<<<
(0,)
VH Array<String>
{'shape': [2, 2], 'array': ['abcd', 'c', '12', 'x12']}
    testarrvh
VH Array<Bool>
[ True False]
VH Bool
True
VH Array<Bool>
[ True]
VH Array<Bool>
[ True False]
bool 1
True
    testarrvh
VH Array<Short>
[-6 -7]
VH Int
-6
VH Array<Int>
[-6]
VH Array<Int>
[-6 -7]
Int -6
-6
Int64 -6
-6
ssize -6
-6
Float -6
-6.0
Complex (-6,0)
(-6+0j)
    testarrvh
VH Array<Short>
[5 6]
VH uInt
5
VH Array<Int>
[5]
VH Array<Int>
[5 6]
Int 5
5
Int64 5
5
ssize 5
5
Float 5
5.0
Complex (5,0)
(5+0j)
    testarrvh
VH Array<Short>
[-16 -17]
VH Int
-16
VH Array<Int>
[-16]
VH Array<Int>
[-16 -17]
Int -16
-16
Int64 -16
-16
ssize -16
-16
Float -16
-16.0
Complex (-16,0)
(-16+0j)
    testarrvh
VH Array<Int>
[15 16]
VH uInt
15
VH Array<Int>
[15]
VH Array<Int>
[15 16]
Int 15
15
Int64 15
15
ssize 15
15
Float 15
15.0
Complex (15,0)
(15+0j)
    testarrvh
VH Array<Int>
[-26 -27]
VH Int
-26
VH Array<Int>
[-26]
VH Array<Int>
[-26 -27]
Int -26
-26
Int64 -26
-26
ssize -26
-26
Float -26
-26.0
Complex (-26,0)
(-26+0j)
    testarrvh
VH Array<uInt>
[25 26]
VH uInt
25
VH Array<Int>
[25]
VH Array<Int>
[25 26]
Int 25
25
Int64 25
25
ssize 25
25
Float 25
25.0
Complex (25,0)
(25+0j)
    testarrvh
VH Array<Int64>
[-36 -37]
VH Int64
-36
VH Array<Int64>
[-36]
VH Array<Int64>
[-36 -37]
Int -36
-36
Int64 -36
-36
ssize -36
-36
Float -36
-36.0
Complex (-36,0)
(-36+0j)
    testarrvh
VH Array<Int64>
[35 36]
VH Int64
35
VH Array<Int64>
[35]
VH Array<Int64>
[35 36]
Int 35
35
Int64 35
35
ssize 35
35
Float 35
35.0
Complex (35,0)
(35+0j)
    testarrvh
VH Array<float>
[-46. -47.]
VH float
-46.0
VH Array<double>
[-46.]
VH Array<double>
[-46. -47.]
Float -46
-46.0
Complex (-46,0)
(-46+0j)
    testarrvh
VH Array<double>
[ 45.  46.]
VH double
45.0
VH Array<double>
[ 45.]
VH Array<double>
[ 45.  46.]
Float 45
45.0
Complex (45,0)
(45+0j)
    testarrvh
VH Array<Complex>
[-56.-66.j -57.-67.j]
VH Complex
(-56-66j)
VH Array<DComplex>
[-56.-66.j]
VH Array<DComplex>
[-56.-66.j -57.-67.j]
Complex (-56,-66)
(-56-66j)
    testarrvh
VH Array<DComplex>
[-76.-86.j -77.-87.j]
VH DComplex
(-76-86j)
VH Array<DComplex>
[-76.-86.j]
VH Array<DComplex>
[-76.-86.j -77.-87.j]
Complex (-76,-86)
(-76-86j)

begin dotest
bool 1
True
bool 0
False
Int -1
-1
Int 10
10
Int64 -123456789013
-123456789013
Int64 123456789014
123456789014
ssize -2
-2
ssize 11
11
Float 3.14
3.1400001049
Float 12
12.0
String this is a string
this is a string
IPos [4, 3, 2]
[2, 3, 4]
IPos [1]
[1]
IPos [2]
[2]
IPos [3]
[3]
VecBool [1, 0, 0, 1]
[True, False, False, True]
VecInt [1, 2, 3, 4]
[1, 2, 3, 4]
VecInt []
[]
VecInt [-1, -2, -3, -4]
[-1, -2, -3, -4]
VecInt [-10]
[-10]
VecInt [10, 11, 12]
[10, 11, 12]
VecInt [1]
[1]
VecComplex [(1,2), (-1,-3), (-1.5,2.5)]
[(1+2j), (-1-3j), (-1.5+2.5j)]
VecStr [a1, a2, b1, b2]
['a1', 'a2', 'b1', 'b2']
VecStr []
[]
VecStr [sc1]
['sc1']
vecbool [0,1]
[False, True]
vecuInt [1,2,4]
[1, 2, 4]
vecuInt []
[]
vecuInt [10]
[10]
vecvecuInt [[1,2,4]]
[[1, 2, 4]]
vecvecuInt []
[]
vecvecuInt []
[]
vecvecuInt [[1],[2],[4]]
[[1], [2], [4]]
vecvecuInt [[20]]
[[20]]
VH Bool
True
VH Int
2
VH Int64
1234567890123
VH double
1.3
VH DComplex
(10-11j)
VH String
str
VH Array<Bool>
[1]
VH Array<Int>
[ 2  4  6  8 10]
VH Array<double>
[ 1.3  4.   5.   6. ]
VH Array<DComplex>
[ 10.-11.j   1. +2.j]
VH Array<String>
['str1', 'str2']
VH Array<String>
{'shape': [2, 2], 'array': ['str1', 'str2', 'str3', 'str4']}
VH Array<String>
{'shape': [2, 2], 'array': ['str1', 'str2', 'str3', 'str4']}
VH Array<String>
{'shape': [2, 2], 'array': ['str1', 'str2', 'str3', 'str4']}
VH Array<DComplex>
(2,)
VH Array<DComplex>
[ 10.-11.j   1. +2.j]
[[2 3]
 [4 5]]
VH Array<Int>
[[2 3]
 [4 5]]
[3 5 7 9]
VH Array<Int>
[3 5 7 9]
[ 3.  5.  7.  9.]
VH Array<double>
[ 3.  5.  7.  9.]
VH Array<double>
[ 3.  5.  7.  9.]
VH Array<DComplex>
[ 20.+10.j]
VH Array<double>
[ 21.]
VH double
21.0
>>>
VH Array<double>
<<<
(0,)
>>>
VH Array<double>
<<<
(1, 0)
vecvh 3
[2, 1.3, array([ True, False], dtype=bool)]
Record int: Int 1
int64: Int64 123456789012
str: String "bc"
vecint: Int array with shape [3]
  [1, 2, 3]

>>>
{'int': 1, 'int64': 123456789012L, 'str': 'bc', 'vecint': array([1, 2, 3], dtype=int32)}
<<<
end dotest

